api/src/extensions/lib/sync-extensions.ts

```diff
@@ -0,0 +1,98 @@
+import { NESTED_EXTENSION_TYPES } from '@directus/extensions';
+import { ensureExtensionDirs } from '@directus/extensions/node';
+import mid from 'node-machine-id';
+import { createWriteStream } from 'node:fs';
+import { mkdir } from 'node:fs/promises';
+import { uptime } from 'node:os';
+import { dirname, join, normalize, resolve } from 'node:path';
+import { pipeline } from 'node:stream/promises';
+import Queue from 'p-queue';
+import { getEnv } from '../../env.js';
+import logger from '../../logger.js';
+import { getMessenger } from '../../messenger.js';
+import { getStorage } from '../../storage/index.js';
+import { getExtensionsPath } from './get-extensions-path.js';
+import { SyncStatus, getSyncStatus, setSyncStatus } from './sync-status.js';
+import { getSyncTimestamp, setSyncTimestamp } from './sync-timestamp.js';
+
+export const syncExtensions = async () => {
+	const messenger = getMessenger();
+
+	const isPrimaryProcess =
+		String(process.env['NODE_APP_INSTANCE']) === '0' || process.env['NODE_APP_INSTANCE'] === undefined;
+
+	const isDone = (await getSyncStatus()) === SyncStatus.DONE;
+
+	const id = await mid.machineId();
+
+	const message = `extensions-sync/${id}`;
+
+	if (isDone === false && isPrimaryProcess === false) {
+		logger.trace('Extensions already being synced to this machine from another process.');
+
+		/**
+		 * Wait until the process that called the lock publishes a message that the syncing is complete
+		 */
+		return new Promise((resolve) => {
+			messenger.subscribe(message, resolve);
+		});
+	}
+
+	const env = getEnv();
+
+	if (!env['EXTENSIONS_LOCATION']) {
+		return;
+	}
+
+	const currentTime = new Date();
+	currentTime.setMilliseconds(0);
+	const systemStartTimestamp = new Date(currentTime.getTime() - uptime() * 1000).toISOString();
+	const isSyncRequired = (await getSyncTimestamp()) !== systemStartTimestamp;
+
+	if (!isSyncRequired) {
+		return;
+	}
+
+	const extensionsPath = getExtensionsPath();
+
+	// Ensure that the local extensions cache path exists
+	await mkdir(extensionsPath, { recursive: true });
+	await setSyncStatus(SyncStatus.SYNCING);
+
+	logger.trace('Syncing extensions from configured storage location...');
+
+	const storage = await getStorage();
+
+	const disk = storage.location(env['EXTENSIONS_LOCATION']);
+
+	// Make sure we don't overload the file handles
+	const queue = new Queue({ concurrency: 1000 });
+
+	for await (const filepath of disk.list(env['EXTENSIONS_PATH'])) {
+		const readStream = await disk.read(filepath);
+
+		// We want files to be stored in the root of `$TEMP_PATH/extensions`, so gotta remove the
+		// extensions path on disk from the start of the file path
+		const destPath = join(
+			extensionsPath,
+			filepath.substring(resolve('/', normalize(env['EXTENSIONS_PATH'])).length - 1)
+		);
```

Comment:
This logic seems flawed as all my extension folders get created without the first character, resulting in them not being recognized as extensions and not loaded.

![image](https://github.com/directus/directus/assets/9389634/0c1c6139-b17d-467d-b46d-289ba134e703)

```suggestion
		const destPath = join(
			extensionsPath,
			filepath.substring(normalize(env['EXTENSIONS_PATH']).length)
		);
```

Reply:
Great catch on the difference in file path conventions! I've just pushed https://github.com/directus/directus/pull/20207/commits/938330116f2a881ce31cd88b64b5bf3746444faa, does it resolve this issue? 🧐

Reply:
This solution does not work im afraid, the underlying issue lies with the use of `resolve()` not with the directory separator. Both directory separators are only 1 character in length but the difference is coming from the Drive letter that is added in `resolve` (hence my suggested change).
![image](https://github.com/directus/directus/assets/9389634/c5729292-3c36-47f9-bdae-a664a3281feb)

Reply:
```suggestion
		const destPath = join(
			extensionsPath,
			filepath.substring(resolve(sep, normalize('./extensions')).length - resolve(sep).length)
		);
```

Reply:
@br41nslug Great idea! Pushed via https://github.com/directus/directus/pull/20207/commits/6e78b0ebea9d9274d000789fc971f071b7096aff. Thank you! 🙏